File: twitter-fetcher/src/test/java/org/jesperancinha/twitter/model/MessageTest.java
package org.jesperancinha.twitter.model;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.jesperancinha.twitter.model.twitter.Message;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class MessageTest {

    private final static Gson gson = new GsonBuilder()
            .setDateFormat("EEE MMM dd HH:mm:ss ZZZZZ yyyy")
            .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
            .create();


    @Test
    public void testEquals_whenEquals_Ok() {
        final Message message1 = gson.fromJson(
                "{\n" +
                        "  \"created_at\": \"Mon Feb 13 12:15:04 +0000 2020\",\n" +
                        "  \"id\": 1216770650121084933,\n" +
                        "  \"id_str\": \"1216770650121084933\",\n" +
                        "  \"text\": \"RT @climalliancech: Congrats @RogerFederer for your declaration #RogerForClimate!\",\n" +
                        "  \"user\": {\n" +
                        "    \"id\": 3024323693,\n" +
                        "    \"name\": \"Fossil Free Penn\"" +
                        "  }" +
                        "}", Message.class);

        final Message message2 = gson.fromJson(
                "{\n" +
                        "  \"created_at\": \"Mon Jan 13 17:15:04 +0000 2020\",\n" +
                        "  \"id\": 1216770650121084933,\n" +
                        "  \"id_str\": \"1216770650121084933\",\n" +
                        "  \"text\": \"RT @Climaaaaaaaate: Congrats @RogerFederer for your declaration #RogerForClimate!\",\n" +
                        "  \"user\": {\n" +
                        "    \"id\": 3024323693,\n" +
                        "    \"name\": \"Fossil Spree Penn\",\n" +
                        "    \"notifications\": null\n" +
                        "  }}", Message.class);

        assertThat(message1).isEqualTo(message2);
    }

    @Test
    public void testHashCode_whenOther_NotSame() {
        final Message message1 = gson.fromJson(
                "{\n" +
                        "  \"created_at\": \"Mon Feb 13 12:15:04 +0000 2020\",\n" +
                        "  \"id\": 1216770650121084933,\n" +
                        "  \"id_str\": \"1216770650121084933\",\n" +
                        "  \"text\": \"RT @climalliancech: Congrats @RogerFederer for your declaration #RogerForClimate!\",\n" +
                        "  \"user\": {\n" +
                        "    \"id\": 3024323693,\n" +
                        "    \"name\": \"Fossil Free Penn\"" +
                        "  }" +
                        "}", Message.class);

        final Message message2 = gson.fromJson(
                "{\n" +
                        "  \"created_at\": \"Mon Jan 13 17:15:04 +0000 2020\",\n" +
                        "  \"id\": 1216770650121084933,\n" +
                        "  \"id_str\": \"1216770650121084933\",\n" +
                        "  \"text\": \"RT @Climaaaaaaaate: Congrats @RogerFederer for your declaration #RogerForClimate!\",\n" +
                        "  \"user\": {\n" +
                        "    \"id\": 3024323693,\n" +
                        "    \"name\": \"Fossil Spree Penn\",\n" +
                        "    \"notifications\": null\n" +
                        "  }}", Message.class);

        assertThat(message1).isNotSameAs(message2);
        assertThat(message1.hashCode()).isEqualTo(message2.hashCode());
    }
}

File: news-cast-explorer-fetcher/src/test/kotlin/org/jesperancinha/newscast/model/twitter/MessageTest.kt
package org.jesperancinha.newscast.model.twitter

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldNotBeSameInstanceAs
import org.junit.jupiter.api.Test

class MessageTest {
    @Test
    @Throws(JsonProcessingException::class)
    fun testEquals_whenEquals_Ok() {
        val message1 = objectMapper.readValue(
            """{
  "created_at": "Mon Feb 13 12:15:04 +0000 2020",
  "id": 1216770650121084933,
  "id_str": "1216770650121084933",
  "text": "Message3",
  "user": {
    "id": 3024323693,
    "name": "Author2"  }}""", Message::class.java)
        val message2 = objectMapper.readValue(
            """{
  "created_at": "Mon Jan 13 17:15:04 +0000 2020",
  "id": 1216770650121084933,
  "id_str": "1216770650121084933",
  "text": "Message5",
  "user": {
    "id": 3024323693,
    "name": "Author3",
    "notifications": null
  }}""", Message::class.java)
        message1 shouldBe message2
    }

    @Test
    @Throws(JsonProcessingException::class)
    fun testHashCode_whenOther_NotSame() {
        val message1 = objectMapper.readValue(
            """{
  "created_at": "Mon Feb 13 12:15:04 +0000 2020",
  "id": 1216770650121084933,
  "id_str": "1216770650121084933",
  "text": "Message3",
  "user": {
    "id": 3024323693,
    "name": "Author2"  }}""", Message::class.java)
        val message2 = objectMapper.readValue(
            """{
  "created_at": "Mon Jan 13 17:15:04 +0000 2020",
  "id": 1216770650121084933,
  "id_str": "1216770650121084933",
  "text": "Message5",
  "user": {
    "id": 3024323693,
    "name": "Author3",
    "notifications": null
  }}""", Message::class.java)
        message1 shouldNotBeSameInstanceAs message2
        message1.hashCode() shouldBe message2.hashCode()
    }

    companion object {
        private val objectMapper = ObjectMapper()
    }
}